### PR TITLE
32 bit compile issues, Makefile envirnomental control variables

### PR DESCRIPTION
**Pull request by nrdmao**
- changes are required to get bess to compile with a 32 bit compiler.
- changes to the Makefile to allow compilation for an alternative architecture.
- changes to the Makefile to allow the option to link against shared libraries.

### CODE REVIEW

File: core/utils/checksum.h [hunk 1]
```diff
@@ -74,6 +74,7 @@ static inline uint32_t CalculateSum(const void *buf, size_t len) {
   }
 #endif
 
+#if __x86_64
```
Comment (sangjinhan) [inline core/utils/checksum.h]: Since this code block doesn't have anything specific to x86, can we use something else (e.g., `__LP64__`)..? While BESS only supports i686 and x86_64, it would be good to keep the code architecture neutral as much as we can.

File: core/modules/exact_match.h [hunk 2]
```diff
@@ -64,7 +64,7 @@ class em_hash {
     }
     return init_val;
 #else
-    return rte_hash_crc(&key, key.key_len, init_val);
+    return rte_hash_crc(&key, len_, init_val);
```
Comment (sangjinhan) [inline core/modules/exact_match.h]: Good point 👍

**Comment by sangjinhan** [comment]
Thank you for the patches! I'll add a 32-bit build test to Travis to keep the code compatible with i686.

File: core/utils/checksum.h [hunk 3]
```diff
@@ -109,9 +110,11 @@ static inline uint32_t CalculateSum(const void *buf, size_t len) {
   // Reduce 64-bit unsigned integer to 32-bit unsigned integer
   // Carry may happens, but no need to complete reduce here
   sum64 = (sum64 >> 32) + (sum64 & 0xFFFFFFFF);
+#endif
 
   // Repeat 16-bit one's complement sum (at sum64)
-  const uint16_t *buf16 = reinterpret_cast<const uint16_t *>(buf64);
+  typedef uint16_t __attribute__((__may_alias__)) u16_p;
```
Comment (sangjinhan) [inline core/utils/checksum.h]: Why did you add this line? My test with g++ 4.8 / 5.4 does not complain about aliasing here...

Comment (nrdmao) [inline core/utils/checksum.h]: Have you run the unit tests? It was not a compile time issue but a run time issue. It seems when there is optimization -O2 or greater the optimizer presumes strict aliasing. Note that this is the same technique used rte_raw_cksum(). I suspect when the remainder of the function is ifdef'd out (as is the case in 32 bit mode) the optimizer kicks in and the strict aliasing issue shows up (I don't think you see the optimizer effects on this section of the code in 64 bit mode).

File: core/utils/checksum.h (hunk 1 above)
Comment (nrdmao) [inline core/utils/checksum.h]: Will do.

Comment (nrdmao) [inline core/utils/checksum.h]: Actually on second look, this code block is all x86 assembly code so it seems like x86_64 is the right thing to do, unless I misunderstand you.

Comment (sangjinhan) [inline core/utils/checksum.h]: Oh you are right. What the heck I was thinking about...

File: core/utils/checksum.h (hunk 3 above)
Comment (sangjinhan) [inline core/utils/checksum.h]: I see. What I did was just build test (with `-m32` option to g++)i, as I don't have a 32-bit env to do runtime test. While it's entirely cosmetic, perhaps we can use `[[gnu:may_alias]]` instead?

**Comment by sangjinhan** [comment]
Please add your name to CONTRIBUTORS.md :)

**Comment by sangjinhan** [comment]
Thanks!

gxldem8kxnffpy9mq5bdgun4o1_1280
